src/System.Text.Primitives/System/Text/Parsing/InvariantUtf8/InvariantUtf8_UInt32.cs

```diff
@@ -151,35 +176,425 @@ public unsafe static bool TryParseUInt32(byte* text, int length, out uint value,
+                        if (nextDigit > 9)
+                        {
+                            bytesConsumed = index;
+                            return true;
+                        }
+                        value = value * 10 + nextDigit;
+                    }
+                    for (int index = UINT_OVERFLOW_LENGTH - 1; index < text.Length; index++)
+                    {
+                        uint nextDigit = text[index] - 48u; // '0'
+                        if (nextDigit > 9)
+                        {
+                            bytesConsumed = index;
+                            return true;
+                        }
+                        // uint.MaxValue is a constant 4294967295
+                        // uint.MaxValue / 10 is 429496729 (integer division truncated)
+                        // (value > uint.MaxValue / 10) implies overflow when appended to; e.g., 4294967300 > 4294967295
+                        // if value == 429496729, any nextDigit greater than 5 implies overflow
+                        if (value > uint.MaxValue / 10 || (value == uint.MaxValue / 10 && nextDigit > 5))
+                        {
+                            bytesConsumed = 0;
+                            value = default(uint);
+                            return false;
+                        }
+                        value = value * 10 + nextDigit;
+                    }
+                }
 
+                bytesConsumed = text.Length;
+                return true;
+            }
+            
             public static partial class Hex
             {
                 public unsafe static bool TryParseUInt32(byte* text, int length, out uint value)
                 {
-                    int consumed;
-                    var span = new ReadOnlySpan<byte>(text, length);
-                    return PrimitiveParser.TryParseUInt32(span, out value, out consumed, EncodingData.InvariantUtf8, 'X');
+                    if (length < 1)
+                    {
+                        value = default(uint);
+                        return false;
+                    }
+
+                    // Parse the first digit separately. If invalid here, we need to return false.
+                    byte firstByte = text[0];
+                    int firstDigit = hexLookup[firstByte];
+                    if (firstDigit == -1)
+                    {
+                        value = default(uint);
+                        return false;
+                    }
+                    value = (uint)firstDigit;
+
+                    if (length <= UINT_OVERFLOW_LENGTH_HEX)
+                    {
+                        // Length is less than or equal to OVERFLOW_LENGTH; overflow is not possible
+                        for (int index = 1; index < length; index++)
+                        {
+                            byte nextByte = text[index];
+                            int nextDigit = hexLookup[nextByte];
```

Comment:
Note that the access to static variables come with static constructor checks - this line basically translates to:

```
if (has cctor run)
    Run static constructor
fetch hexLookup
```

It is often beneficial to cache the static variable in a local, to guarantee that the static cctor check is done just once.